src/ucs/stats/stats.c

```diff
@@ -477,95 +481,108 @@ void ucs_stats_node_free(ucs_stats_node_t *node)
-                    (ucs_stats_context.aggrgt_sum_db_alloc_size *
-                     sizeof(*ucs_stats_context.aggrgt_sum_db));
-            ucs_stats_context.aggrgt_sum_db = ucs_realloc(
-                    ucs_stats_context.aggrgt_sum_db, total_alloc_size_bytes,
-                    (const char*)UCS_STATS_CLSID_TO_NAME_DB_NAME);
+        status = ucs_array_reserve(aggrgt_sum_db_array_t,
+                                   &ucs_stats_context.aggrgt_sum_db,
+                                   new_size);
+        if (status != UCS_OK) {
+            return status;
         }
 
-        /* Initialize entry */
-        ucs_snprintf_safe(
-                ucs_stats_context
-                        .aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
-                        .class_name,
-                sizeof(ucs_stats_context
-                               .aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
-                               .class_name),
-                "%s", class_name);
-        ucs_stats_context.aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
-                .aggregate_sum_class_offset =
-                ucs_stats_context.aggrgt_sum_db_last_offset;
-
-        /* Initialize return value */
-        offset = ucs_stats_context.aggrgt_sum_db_last_offset;
-
-        /* Return class_id */
-        *class_id = ucs_stats_context.aggrgt_sum_db_size;
-
-        /* Update global offset */
-        ucs_stats_context.aggrgt_sum_db_last_offset += num_counters_in_class;
-
-        /* Reallocate counters names db */
-        ucs_stats_context.aggrgt_cnt_names_db_size += num_counters_in_class;
-        total_alloc_size_bytes =
-                (ucs_stats_context.aggrgt_cnt_names_db_size *
-                 sizeof(*ucs_stats_context.aggrgt_cnt_names_db));
-        ucs_stats_context.aggrgt_cnt_names_db = ucs_realloc(
-                ucs_stats_context.aggrgt_cnt_names_db, total_alloc_size_bytes,
-                (const char*)UCS_STATS_AGGRGT_SUM_CNT_NAMES_DB);
-
-        cnt_agrgt_sum_index = offset;
-        ucs_for_each_bit(counter_index, filter_node->counters_bitmask) {
-            ucs_list_for_each(temp_node, &filter_node->type_list_head,
-                              type_list) {
-                ucs_stats_context.aggrgt_cnt_names_db[cnt_agrgt_sum_index].name =
-                        node->cls->counter_names[counter_index];
-                ucs_stats_context.aggrgt_cnt_names_db[cnt_agrgt_sum_index]
-                        .class_name = node->cls->name;
-            }
-            cnt_agrgt_sum_index++;
+        capacity = ucs_array_capacity(&ucs_stats_context.aggrgt_sum_db);
+        ucs_array_set_length(&ucs_stats_context.aggrgt_sum_db, capacity);
+    }
+
+    element = &ucs_array_elem(&ucs_stats_context.aggrgt_sum_db, db_size);
+    printf("element = %p\n", element);
+
+    /* Initialize entry */
+    ucs_snprintf_safe(element->class_name, sizeof(element->class_name),
+                      "%s", class_name);
```

Comment:
element->class name is not really used
you use aggrgt_cnt_names_db which is a different array

Reply:
Please review the code in test_stats.cc (example code for reading the counters) - class_name is indeed used by the user.

    void
    read_and_check_aggrgt_sum_stats(ucs_stats_node_t *data_nodes[NUM_DATA_NODES])
    {
        const ucs_stats_aggrgt_sum_counter_name_t *aggregate_cnt_names_db = NULL;
        const ucs_stats_aggrgt_sum_clsid_to_name_db_t *clsid_to_name_db = NULL;
        ucs_stats_counter_t ucs_stats_aggregate_sum_counters
                [UCS_STATS_NUM_AGGREGATE_SUM_COUNTERS_MAX];
        size_t size;
        size_t n_counters;
        uint32_t i;

        /* Test */
        n_counters = ucs_stats_aggregate(ucs_stats_aggregate_sum_counters,
                                         UCS_STATS_NUM_AGGREGATE_SUM_COUNTERS_MAX,
                                         &clsid_to_name_db);

        ucs_stats_aggregate_get_counter_names(&aggregate_cnt_names_db, &size);

        ASSERT_NE(0ul, n_counters);
        ASSERT_FALSE(aggregate_cnt_names_db == NULL);

        /* Example for processing the statistics */
        for (i = 0; i < n_counters; i++) {
            if (aggregate_cnt_names_db[i].counter_name && aggregate_cnt_names_db[i].class_name) {
               ucs_print("[%u] pid=%d | CNT: cls_name=<%s> Name=%s, value=%lu\n",
                         i,
                         getpid(),
                         aggregate_cnt_names_db[i].class_name,
                         aggregate_cnt_names_db[i].counter_name,
                         ucs_stats_aggregate_sum_counters[i]);
           }
       }
    }

Reply:
@shuki-zanyovka the test is using class_name from ucs_stats_aggrgt_sum_counter_name_t . 
i am talking about removing class_name from ucs_stats_aggrgt_sum_clsid_to_name_db_t.
you have 2 structs with class_name field. only one of them is used.